base/dates/periods.jl

```diff
@@ -166,10 +166,44 @@ type CompoundPeriod <: AbstractTime
                     break
                 end
             end
-            return new(reverse!(pc))
+            p = reverse(pc)
+            n = length(p)
         else
             return new(resize!(p, n))
         end
+        # canonicalize Period values such that they are all either
+        # positive or negative.
+        if n > 0
+            # Since Periods have been reduced we know that the largest
+            # Period type will indicate the sign.
+            s = sign(value(p[1]))
+
+            pc = sizehint!(Period[], n)
+            P = typeof(p[n])
+            v = value(p[n])
+            i = n - 1
+            while true
+                Pc, f = coarserperiod(P)
+                if i > 0 && typeof(p[i]) == P
```

Comment:
why should this branch ever be taken?  shouldn't the previous loop have already coalesced equal period types?